### ISSUE TITLE
Invoking garbage collection globally

### ISSUE THREAD

**Issue by kapoorlab** [napari/napari#5543]
## 🚀 Feature
In order to keep the memory free for computations it would help if Garbage collection is invoked globally in Napari. It could be triggered after a layer is popped programmatically or manually using the delete layer button.

## Motivation

When analyzing data with the plugins the users want to change the data and start another analysis or re do a calculation in which case the program may replace the layers with new layers and it would lead to increase in the memory usage if the garbage collection is not called in the plugin or globally. Plugin developers can also put this garbage collection in their plugins however even doing this globally would be useful. 

## Pitch

Invoking garbage collection linked to poping a layer by either pressing the Delete button from the Napari UI or if a layer is removed programmatically using the Napari terminal or by the plugin in use.

## Alternatives

Calling gc.collect() in the plugin using qtimer to allow for a delay in freeing the memory.

**Comment by jni** [comment]
I'm not immediately opposed to this, but there's probably performance ramifications to calling collect() too often. We'd need to explore the tradeoffs.

**Comment by kapoorlab** [comment]
How about giving this choice to the users? I was thinking of making the bin widget grow in thickness as the heap grows and having a keyboard-mouse shortcut that when invoked on the bin starts garbage collection and the bin widget visibly shrinks/become normal size once this action is take. Currently I am using the Napari console to do this after I pop images from the viewer.

**Comment by Czaki** [comment]
I do not think that call `gc.collect` after removing the layer could be problematic.

**Comment by Carreau** [comment]
> I do not think that call `gc.collect` after removing the layer could be problematic.

I think it can be; if always triggered then when you remove a bunch of layers in succession
it will  trigger many gc passes, that can make napari unusually slow.

We had the issue in test where we were calling gc.collect too many time in test teardown making the test unbearably slow.

In general my experience is that `gc.collect()` is good enough and already has it's own logic on when to collect that trying to add logic on top of this to make it more efficient just make things worse, or delay the inevitable.

I'm not opposed to adding gc.collect, but I would advise against it.

**Comment by Czaki** [comment]
> I think it can be; if always triggered then when you remove a bunch of layers in succession
> it will trigger many gc passes, that can make napari unusually slow.

Calling `gc.collect()` on `delete` event will be ineffective, as the event will store a reference to layer. 

My idea at this moment is to use Qt timers for this. and Have single QTimer in MainWindow object that will be started in close event handler. If it is already started then it will be restarted. 

Such an approach will guarantee `gc.collect()` call is enough fast that memory will be free before the user manually loads a new file. Also using a timer guarantee that independently of the number of removed layers `gc.collect` will be called once.
